src/main/java/net/fabricmc/loader/metadata/V0ModMetadataParser.java

```diff
@@ -0,0 +1,430 @@
+
+final class V0ModMetadataParser {
+	private static final Pattern WEBSITE_PATTERN = Pattern.compile("\\((.+)\\)");
+	private static final Pattern EMAIL_PATTERN = Pattern.compile("<(.+)>");
+
+	public static LoaderModMetadata parse(JsonReader reader) throws IOException, ParseMetadataException {
+		// All the values the `fabric.mod.json` may contain:
+		// Required
+		String id = null;
+		Version version = null;
+
+		// Optional (mod loading)
+		Map<String, ModDependency> requires = new HashMap<>();
+		Map<String, ModDependency> conflicts = new HashMap<>();
+		V0ModMetadata.Mixins mixins = null;
+		ModEnvironment environment = ModEnvironment.UNIVERSAL; // Default is always universal
+		String initializer = null;
+		List<String> initializers = new ArrayList<>();
+
+		String name = null;
+		String description = null;
+		Map<String, ModDependency> recommends = new HashMap<>();
+		List<Person> authors = new ArrayList<>();
+		List<Person> contributors = new ArrayList<>();
+		ContactInformation links = null;
+		String license = null;
+
+		while (reader.hasNext()) {
+			switch (reader.nextName()) {
+			case "schemaVersion":
+				// Duplicate field, make sure it matches our current schema version
+				if (reader.peek() != JsonToken.NUMBER) {
+					throw new ParseMetadataException("Duplicate \"schemaVersion\" field is not a number", reader);
+				}
+
+				final int read = reader.nextInt();
+
+				if (read != 0) {
+					throw new ParseMetadataException(String.format("Duplicate \"schemaVersion\" field does not match the predicted schema version of 0. Duplicate field value is %s", read), reader);
+				}
+
+				break;
+			case "id":
+				if (reader.peek() != JsonToken.STRING) {
+					throw new ParseMetadataException("Mod id must be a non-empty string with a length of 3-64 characters.", reader);
+				}
+
+				id = reader.nextString();
+				break;
+			case "version":
+				if (reader.peek() != JsonToken.STRING) {
+					throw new ParseMetadataException("Version must be a non-empty string", reader);
+				}
+
+				try {
+					version = VersionDeserializer.deserialize(reader.nextString());
+				} catch (VersionParsingException e) {
+					throw new ParseMetadataException("Failed to parse version", e);
+				}
+
+				break;
+			case "requires":
+				V0ModMetadataParser.readDependenciesContainer(reader, requires, "requires");
+				break;
+			case "conflicts":
+				V0ModMetadataParser.readDependenciesContainer(reader, conflicts, "conflicts");
+				break;
+			case "mixins":
+				mixins = V0ModMetadataParser.readMixins(reader);
+				break;
+			case "side":
+				if (reader.peek() != JsonToken.STRING) {
+					throw new ParseMetadataException("Side must be a string", reader);
+				}
+
+				switch (reader.nextString()) {
+				case "universal":
+					environment = ModEnvironment.UNIVERSAL;
+					break;
+				case "client":
+					environment = ModEnvironment.CLIENT;
+					break;
+				case "server":
+					environment = ModEnvironment.SERVER;
+					break;
+				}
```

Comment:
warn for default case

Reply:
I think the default branch of the switch block should be to fail, you don't exactly have a 4th mod environment